File: services/language-service/src/main/java/dev/animedia/languageservice/application/service/SearchLanguageService.java
package dev.animedia.languageservice.application.service;

import dev.animedia.languageservice.application.dto.LanguageDto;
import dev.animedia.languageservice.application.dto.SearchLanguageDto;
import dev.animedia.languageservice.application.mapper.LanguageApplicationMapper;
import dev.animedia.languageservice.application.usecase.SearchLanguageUseCase;
import dev.animedia.languageservice.domain.model.Language;
import dev.animedia.languageservice.domain.model.Page;
import dev.animedia.languageservice.domain.repository.LanguageQueryRepository;

public class SearchLanguageService implements SearchLanguageUseCase {
    private final LanguageQueryRepository languageQueryRepository;
    private final LanguageApplicationMapper languageApplicationMapper;

    public SearchLanguageService(
        LanguageQueryRepository languageQueryRepository,
        LanguageApplicationMapper languageApplicationMapper
    ) {
        this.languageQueryRepository = languageQueryRepository;
        this.languageApplicationMapper = languageApplicationMapper;
    }

    @Override
    public Page<LanguageDto> search(SearchLanguageDto searchDto) {
        Page<Language> languagePage = languageQueryRepository.search(searchDto.codes(), searchDto.names(), searchDto.isActive(), searchDto.pagination());
        var content = languagePage.content().stream().map(languageApplicationMapper::toDto).toList();
        return new Page<>(
            content,
            languagePage.totalElements(),
            languagePage.totalPages(),
            languagePage.pageNumber(),
            languagePage.pageSize(),
            languagePage.isFirst(),
            languagePage.isLast(),
            languagePage.hasNext(),
            languagePage.hasPrevious()
        );
    }
}

File: services/social-service/src/core/Application/Language/Service/Private/SearchLanguageService.php
<?php

declare(strict_types=1);

namespace Core\Application\Language\Service\Private;

use Core\Application\Language\DTO\Public\LanguageResponseDto;
use Core\Application\Language\Mapper\LanguageApplicationMapperInterface;
use Core\Application\Language\UseCase\Private\SearchLanguageUseCase;
use Core\Domain\Language\Repository\LanguageQueryRepositoryInterface;
use Core\Domain\Shared\Pagination\Entity\Page;
use Core\Domain\Shared\Pagination\Entity\Pageable;

final readonly class SearchLanguageService implements SearchLanguageUseCase
{
    public function __construct(
        private LanguageQueryRepositoryInterface $languageQueryRepository,
        private LanguageApplicationMapperInterface $languageApplicationMapper
    ) {}

    /**
     * @return Page<LanguageResponseDto>
     */
    #[\Override]
    public function execute(?string $languageIsoCode, ?string $name, ?bool $isActive, Pageable $pageable): Page
    {
        $languagePage = $this->languageQueryRepository->search(
            pageable: $pageable,
            active: $isActive,
            languageIsoCode: $languageIsoCode,
            name: $name
        );

        $languageDtoList = [];
        foreach ($languagePage->content as $languageEntity) {
            $languageDtoList[] = $this->languageApplicationMapper->toPrivateLanguageDto($languageEntity);
        }

        return $languagePage->changeContent($languageDtoList);
    }
}
